[ci-skip] Improve error-reporting guide (#46094)
rails/rails: guides/source/error_reporting.md

After reading this guide, you will know:

* How to use Rails' error reporter to capture and report errors.

(Because error reporting libraries are mainly for production use, this guide is mostly for production environment too.

> However, you may want to test related setup in development by temporarily activating the libraries locally.)

Author: Removed this blurb as it's confusing. The Rails error reporter works in every environment, but this made it seem like it only works in production. It's really unnecessary.

Reviewer: I asked for this section. Yes you can configure error reporters in all environments, but the use case is mostly in production or staging to report to services such as Sentry &co.

Author: Okay. Can I move it to the point where we mention error services, though? I think it's more relevant there. I found it really confusing up here and had to double check whether the Rails error reporter cared about my environment, even though I had followed the implementation. 😅

Author: Added:

> Note: The Rails error-reporter will always call registered subscribers, regardless of your environment. However, many error-reporting services only report errors in production by default. You should configure and test your setup across environments as needed.

```ruby
begin
  do_something
rescue SomethingIsBroken => error
  MyErrorReportingService.notify(error)
end
```

with a consistent interface:

```ruby
Rails.error.handle(SomethingIsBroken) do
  do_something
end
```

Rails wraps all executions (such as HTTP requests, jobs, and `rails runner` invocations) in the error reporter, so any unhandled errors raised in your app will automatically be reported to your error-reporting service via their subscribers.

This means that third-party error-reporting libraries no longer need to insert a Rack middleware or do any monkey-patching to capture unhandled exceptions. Libraries that use ActiveSupport can also use this to non-intrusively report warnings that would previously have been lost in logs.

Using the Rails' error reporter is not required. All other means of capturing errors still work.

### Subscribing to the Reporter

To use the error reporter, you need a _subscriber_. A subscriber is any object with a `report` method. When an error occurs in your application or is manually reported, the Rails error reporter will call this method with the error object and some options.

> Some error-reporting libraries, such as [Sentry's](https://github.com/getsentry/sentry-ruby/blob/master/sentry-rails/lib/sentry/rails/error_subscriber.rb) and [Honeybadger's](https://docs.honeybadger.io/lib/ruby/integration-guides/rails-exception-tracking/), automatically register a subscriber for you. Consult your provider's documentation for more details.

Reviewer 2: Should we change that sentry URL to point to a specific commit? I'm worried that the file may be moved or whatever which would break the link. 

So that would be https://github.com/getsentry/sentry-ruby/blob/e18ce4b6dcce2ebd37778c1e96164684a1e9ebfc/sentry-rails/lib/sentry/rails/error_subscriber.rb

Author: 👍

end
```

with a consistent interface:

```rb
Rails.error.handle(SomethingIsBroken) do
  do_something
end
```

Rails wraps all executions (such as HTTP requests, jobs, and `rails runner` invocations) in the error reporter, so any unhandled errors raised in your app will automatically be reported to your error-reporting service via their subscribers.

> This means that third-party error-reporting libraries no longer need to insert a Rack middleware or do any monkey-patching to capture unhandled exceptions. ActiveSupport libraries can also use this to non-intrusively report warnings that would previously have been lost in logs.

Reviewer 3: ```suggestion
This means that third-party error-reporting libraries no longer need to insert a Rack middleware or do any monkey-patching to capture unhandled exceptions. Libraries that uses Active Support can also use this to non-intrusively report warnings that would previously have been lost in logs.
```

This means that third-party error-reporting libraries no longer need to insert a Rack middleware or do any monkey-patching to capture unhandled exceptions. ActiveSupport libraries can also use this to non-intrusively report warnings that would previously have been lost in logs.

Using the Rails' error reporter is not required. All other means of capturing errors still work.

* Application or error reporting libraries don't need to insert a Rack middleware to capture unhandled exceptions from requests anymore.
* To ActiveSupport-aware libraries, this can be used to report errors to the host application.
* It decouples application code from error reporting libraries.
* It reduces boilerplate code for handling exceptions.

> * Error reporting libraries will need less monkey-patches and be less intrusive to applications.

Author: Removed the bullet list of "benefits" as they made it look too academic, and didn't share the tone you normally get from the Rails docs. They were also not very distinct, and could be summarized in fewer sentences.

### Subscribing to the Reporter

### Subscribe To The Reporter
To use the error reporter, you need a _subscriber_. A subscriber is any object with a `report` method. When an error occurs in your application or is manually reported, the Rails error reporter will call this method with the error object and some options.

An error subscriber is expected to have a `report` method that takes an exception object and a few options.
For example:
Some error-reporting libraries, such as [Sentry's](https://github.com/getsentry/sentry-ruby/blob/master/sentry-rails/lib/sentry/rails/error_subscriber.rb) and [Honeybadger's](https://docs.honeybadger.io/lib/ruby/integration-guides/rails-exception-tracking/), automatically register a subscriber for you. Consult your provider's documentation for more details.

You may also create a custom subscriber. For example:

```rb

> # config/initializers/error_subscriber.rb

Reviewer 4: I think it's still better to hint people where to put this if they want to add one themselves. This feature is not widely adopted by error-reporting services so many users need to create the subscriber manually.

Author: Yeah. I removed it because I felt it was _too strong_ a hint (ie "it has to be in this file"), but I see your point. I'll restore it.

```

You can register as many subscribers as you wish. Rails will call them in turn, in the order in which they were registered.

### Using the Error Reporter

> There are three ways you can use the error reporter :

Reviewer 4: ```suggestion
There are three ways you can use the error reporter:
```

- `handled`: a `Boolean` to indicate if the error was handled. This is set to `true` by default. `#record` sets this to `false`.
- `severity`: a `Symbol` describing the severity of the error. Expected values are: `:error`, `:warning`, and `:info`. `#handle` sets this to `:warning`, while `#record` sets it to `:error`.
- `context`: a `Hash` to provide more context about the error, like request or user details
- `source`: a `String` about the source of the error. The default source is `"application"`. Errors reported by internal libraries may set other sources; the Redis cache library may use `"redis_cache_store.active_support"`, for instance. Your subscriber can use the source to ignore errors you aren't interested in.

```ruby
Rails.error.handle(context: {user_id: user.id}, severity: :info) do
  # ...
end
```

### Filtering by Error Class

With `Rails.error.handle` and `Rails.error.record`, you can also choose to only report errors of a certain class. For example:

```ruby
Rails.error.handle(IOError) do
  1 + '1' # raises TypeError
end
1 + 1 # TypeErrors are not IOErrors, so this will *not* be executed
```

> Here, the `TypeError` will not be captured by the Rails error reporter. Only instances of  `IOError` will be reported. Any other errors will be raised as normal.

Reviewer 2: ```suggestion
Here, the `TypeError` will not be captured by the Rails error reporter. Only instances of  `IOError` and its descendants will be reported. Any other errors will be raised as normal.
```
?

end
```

Any options you pass will be passed on the error subscribers.

### Error-reporting options

All 3 reporting APIs (`#handle`, `#record`, and `#report`) support the following options, which are then passed along to all registered subscribers:

- `handled`: a `Boolean` to indicate if the error was handled. This is set to `true` by default. `#record` sets this to `false`.
- `severity`: a `Symbol` describing the severity of the error. Expected values are: `:error`, `:warning`, and `:info`. `#handle` sets this to `:warning`, while `#record` sets it to `:error`.
- `context`: a `Hash` to provide more context about the error, like request or user details
- `source`: a `String` about the source of the error. The default source is `"application"`. Errors reported by internal libraries may set other sources; the Redis cache library may use `"redis_cache_store.active_support"`, for instance. Your subscriber can use the source to ignore errors you aren't interested in.

> ```ruby

Reviewer 4: nitpick: can we standardise on using `rb` instead of `ruby`?

Author: Right...my old habits

Rails.error.handle(context: {user_id: user.id}, severity: :info) do
  # ...
end
```

### Filtering by Error Class

With `Rails.error.handle` and `Rails.error.record`, you can also choose to only report errors of a certain class. For example:

> ```ruby

Reviewer: ```
byroot@MacBook-Pro:rails:(main)$ ag -sQ '```ruby'  | wc -l
    2233
byroot@MacBook-Pro:rails:(main)$ ag -sQ '```rb'  | wc -l
      29
```

If anything I think it should be the other way around.

> ```ruby

Reviewer 4: Yeah sure I only compared it with the rest of the document. Now it's a great opportunity to make it consistent.

> ```ruby

Author: Okay...will switch to `ruby` then.

  end
end
```

> When using the Rails error reporter, errors may be reported multiple times. For instance, calling `Rails.error.record` will report the error via your subscriber, then raise it. If you have other error mechanisms like a Rack middleware, they will then capture the exception and report it again, then raise it once more, where it may go through your subscriber a second time. 

Author: @casperisfine this was something I noticed lately. Since the executor calls `Rails.error.report`, if I call `Rails.error.record` in my controller, the error will be reported twice? Do you think this is worth fixing?  FWIW, the fix would probably be the approach I described below of adding an instance variable, but doing it in the executor. I can send in a PR for that.

Reviewer: > if I call `Rails.error.record` in my controller, the error will be reported twice?

Why would you do that though? The point of `record` is when you know the exception will be handled by the caller at some point.

Reviewer 4: The user may add more context in the controller's `Rails.error.record`, so the error captured there will have more context than when being captured by the executor. Sentry also has a similar mechanism to workaround the double-reporting issue.

Reviewer: > The user may add more context in the controller's Rails.error.record

Interesting, I didn't think of that. I'll think a bit more about it, but yeah, it may make sense to dedup in that case.

  end
end
```

When using the Rails error reporter, errors may be reported multiple times. For instance, calling `Rails.error.record` will report the error via your subscriber, then raise it. If you have other error mechanisms like a Rack middleware, they will then capture the exception and report it again, then raise it once more, where it may go through your subscriber a second time.

You should adjust your report functionality so it skips reporting an exception it has seen before. One way to do this is to set a custom instance variable on the exception object:

```ruby
module MySdk
  class Reporter
    def report_error(e, **options)
      return if e.instance_variable_get(:@___myservice_reported)
    

>       e.instance_variable_set(:@___myservice_reported, true)

Reviewer 2: I think I'll integrate a similar mechanism, so that we don't have to suggest that kind of stuff.
